appinventor/components/src/com/google/appinventor/components/runtime/util/JavaReplaceWithMappings.java

```diff
@@ -0,0 +1,189 @@
+// Copyright 2020  MIT, All rights reserve
+// Released under the Apache License, Version 2.0
+// http://www.apache.org/licenses/LICENSE-2.0
+
+package com.google.appinventor.components.runtime.util;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Java implementation of replace-with-mappings. Used for ease
+ * of reasoning about the solution.
+ * See runtime.scm
+ * <p>
+ * TODO: Might be better to re-implement this in Scheme/Kawa in the future.
+ * TODO: Add unit tests
+ */
+public final class JavaReplaceWithMappings {
+  /**
+   * Replaces the specified text string with the specified mappings,
+   * which is a map containing Key Value pairs where the key is the
+   * substring to replace, and the value is the value to replace the substring with.
+   * A mode parameter is also specified to indicate the order of mapping application:
+   * 0 - longest string first
+   * 1 - dictionary order
+   * 2 - earliest string first
+   *
+   * @param text     Text to apply mappings to
+   * @param mappings Map containing mappings
+   * @param mode     Mode to use for replacing mappings
+   * @return Text with the mappings applied
+   */
+  public static String replaceWithMappings(String text, Map<Object, Object> mappings, int mode) {
+    // Iterate over all the mappings
+    Iterator<Map.Entry<Object, Object>> it = mappings.entrySet().iterator();
+
+    // Construct a new map for <String, String> mappings in order to support
+    // look-ups for non-pure String values (e.g. numbers)
+    Map<String, String> stringMappings = new HashMap<>();
+
+    // Construct a new List to store the Map's keys
+    List<String> keys = new ArrayList<>();
+
+    while (it.hasNext()) {
+      Map.Entry<Object, Object> current = it.next();
+
+      // Get Key & Value, and update string mappings map
+      String key = current.getKey().toString();
+      String value = current.getValue().toString();
+      stringMappings.put(key, value);
+
+      // Add key
+      keys.add(key);
+    }
+
+    // TODO: Should probably refactor this somehow so it's less hardcoded.
+    // TODO: Maybe create new function defs in scheme instead, and map mode -> function call?
+    // Note: We do not check mode: 1 since that's the default order
+    // that the keys were inserted in (the dictionary order)
+    if (mode == 0) {
+      sortKeysOnLargestSizeFirst(keys);
+    } else if (mode == 2) {
+      sortKeysOnEarliestOccurrenceFirst(text, keys);
+    }
+
+    // We will construct a union regex pattern from the keys
+    StringBuilder patternBuilder = new StringBuilder();
+
+    for (int i = 0; i < keys.size(); ++i) {
+      String key = keys.get(i);
+
+      // Append mapping that we want to replace to the regex pattern
+      patternBuilder.append(key);
```

Comment:
Probably need to escape any `|` in key as `\\|`, otherwise they will be incorporated into the regex.

Reply:
For the replacement itself, it should not affect the mapping itself, since we check for the key's presence before we apply the mapping when we encounter a regex match. However, that is a very fair point, since it can potentially make the regex slower (and not entirely correct). I will update to escape the keys.